File: android-kit-base/src/test/java/com/mparticle/kits/FilteredApiRequestTest.java
package com.mparticle.kits;

import com.mparticle.MParticle;
import com.mparticle.identity.IdentityApiRequest;

import org.junit.Test;
import org.mockito.Mockito;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;

public class FilteredApiRequestTest {

    @Test
    public void testEmptyApiRequest() {
        IdentityApiRequest request = IdentityApiRequest.withEmptyUser().build();
        KitIntegration mockIntegration = Mockito.mock(KitIntegration.class);
        KitConfiguration mockConfiguration = Mockito.mock(KitConfiguration.class);
        Mockito.when(mockConfiguration.shouldSetIdentity(Mockito.any(MParticle.IdentityType.class))).thenReturn(true);
        Mockito.when(mockIntegration.getConfiguration()).thenReturn(mockConfiguration);
        FilteredIdentityApiRequest filteredRequest = new FilteredIdentityApiRequest(request, mockIntegration);

        assertEquals(0, filteredRequest.userIdentities.size());
    }

    @Test
    public void testApiRequestFilter() {
        IdentityApiRequest request = IdentityApiRequest.withEmptyUser()
                .email("testEmail")
                .customerId("1234")
                .userIdentity(MParticle.IdentityType.Alias, "alias")
                .userIdentity(MParticle.IdentityType.Google, "mparticle@gmail")
                .build();
        KitIntegration mockIntegration = Mockito.mock(KitIntegration.class);
        KitConfiguration mockConfiguration = Mockito.mock(KitConfiguration.class);
        Mockito.when(mockConfiguration.shouldSetIdentity(MParticle.IdentityType.Email)).thenReturn(true);
        Mockito.when(mockConfiguration.shouldSetIdentity(MParticle.IdentityType.Alias)).thenReturn(true);
        Mockito.when(mockConfiguration.shouldSetIdentity(MParticle.IdentityType.CustomerId)).thenReturn(false);
        Mockito.when(mockConfiguration.shouldSetIdentity(MParticle.IdentityType.Google)).thenReturn(false);
        Mockito.when(mockIntegration.getConfiguration()).thenReturn(mockConfiguration);
        FilteredIdentityApiRequest filteredRequest = new FilteredIdentityApiRequest(request, mockIntegration);

        assertEquals(4, filteredRequest.userIdentities.size());
        assertEquals(2, filteredRequest.getUserIdentities().size());
    }
}

File: android-kit-base/src/test/kotlin/com/mparticle/kits/FilteredApiRequestTest.kt
package com.mparticle.kits

import com.mparticle.MParticle.IdentityType
import com.mparticle.identity.IdentityApiRequest
import org.junit.Assert
import org.junit.Test
import org.mockito.Mockito

class FilteredApiRequestTest {
    @Test
    fun testEmptyApiRequest() {
        val request = IdentityApiRequest.withEmptyUser().build()
        val mockIntegration = Mockito.mock(
            KitIntegration::class.java
        )
        val configMock = Mockito.mock(KitConfiguration::class.java)
        mockIntegration.setConfiguration(configMock)
        Mockito.`when`(
            configMock.shouldSetIdentity(
                anyObject(
                    IdentityType::class.java
                )
            )
        ).thenReturn(true)
        val filteredRequest = FilteredIdentityApiRequest(request, mockIntegration)
        Assert.assertEquals(0, filteredRequest.userIdentities.size.toLong())
    }

    @Test
    fun testApiRequestFilter() {
        val request = IdentityApiRequest.withEmptyUser()
            .email("testEmail")
            .customerId("1234")
            .userIdentity(IdentityType.Alias, "alias")
            .userIdentity(IdentityType.Google, "mparticle@gmail")
            .build()
        val mockIntegration = Mockito.mock(
            KitIntegration::class.java
        )
        val mockConfiguration = Mockito.mock(
            KitConfiguration::class.java
        )
        Mockito.`when`(mockConfiguration.shouldSetIdentity(IdentityType.Email)).thenReturn(true)
        Mockito.`when`(mockConfiguration.shouldSetIdentity(IdentityType.Alias)).thenReturn(true)
        Mockito.`when`(mockConfiguration.shouldSetIdentity(IdentityType.CustomerId))
            .thenReturn(false)
        Mockito.`when`(mockConfiguration.shouldSetIdentity(IdentityType.Google)).thenReturn(false)
        mockIntegration.setConfiguration(mockConfiguration)
        val filteredRequest = FilteredIdentityApiRequest(request, mockIntegration)
        Assert.assertEquals(4, filteredRequest.userIdentities.size.toLong())
        Assert.assertEquals(2, filteredRequest.getUserIdentities().size.toLong())
    }
}
